src/Interface/HasOrder.agda

```diff
@@ -2,31 +2,142 @@
 
 module Interface.HasOrder where
 
+open import Prelude hiding (_⊔_; suc; isPartialOrder; isEquivalence; trans)
 open import Level using (_⊔_; suc)
 open import Relation.Binary using (Rel)
-open import Relation.Binary.Definitions using (Decidable; Antisymmetric)
-open import Relation.Binary.Structures using (IsPreorder; IsPartialOrder)
+open import Relation.Binary.Definitions using (Decidable; Antisymmetric; Irreflexive; Asymmetric)
+open import Relation.Binary.Structures using (IsPreorder; IsPartialOrder; IsStrictPartialOrder; IsEquivalence)
+open import Relation.Binary.Structures using (IsDecTotalOrder ; IsStrictTotalOrder; IsTotalOrder)
 
-module _ {a ℓ ℓ₂} (A : Set a) (_≈_ : Rel A ℓ) where
+module _ {α ℓ} (A : Set α) (_≈_ : Rel A ℓ) where
 
-  record HasPreorder : Set (a ⊔ ℓ ⊔ suc ℓ₂) where
-    infix 4 _≤_
+  record HasPreorder : Set (α ⊔ suc ℓ) where
+    infix 4 _≤_ _<_
     field
-      _≤_    : Rel A ℓ₂
-      isPreorder : IsPreorder _≈_ _≤_
+      _≤_ _<_       : Rel A ℓ
+      ≤-isPreorder  : IsPreorder _≈_ _≤_
+      <-irrefl      : Irreflexive _≈_ _<_
+      ≤⇔<∨≈         : ∀ {a b} → a ≤ b ⇔ (a < b ⊎ a ≈ b)
 
-  record HasPartialOrder : Set (a ⊔ ℓ ⊔ suc ℓ₂) where
-    open HasPreorder
+  record HasPartialOrder : Set (α ⊔ suc ℓ) where
     field
       hasPreorder  : HasPreorder
-      antisym      : Antisymmetric _≈_ (_≤_ hasPreorder)
+      ≤-antisym    : Antisymmetric _≈_ (HasPreorder._≤_ hasPreorder)
 
-  record HasDecPartialOrder : Set (a ⊔ ℓ ⊔ suc ℓ₂) where
+    open HasPreorder hasPreorder
+
+    <⇒≤∧≠ : ∀{x y} → x < y → x ≤ y × ¬ (x ≈ y)
+    <⇒≤∧≠ {x}{y} x<y = (Equivalence.from ≤⇔<∨≈ (inj₁ x<y)) , λ x≈y → <-irrefl x≈y x<y
+
+    ≤-antisym⇒<-asym : Antisymmetric _≈_ _≤_ → Asymmetric _<_
+    ≤-antisym⇒<-asym antisym x<y y<x = proj₂ (<⇒≤∧≠ x<y) (antisym (proj₁ (<⇒≤∧≠ x<y)) (proj₁ (<⇒≤∧≠ y<x)))
+
+    <-asymmetric : Asymmetric _<_
+    <-asymmetric = ≤-antisym⇒<-asym ≤-antisym
+
+    <⇒¬>⊎≈ : ∀{x y} → x < y → ¬ (y < x ⊎ y ≈ x)
+    <⇒¬>⊎≈ x<y (inj₁ y<x) = <-asymmetric x<y y<x
+    <⇒¬>⊎≈ x<y (inj₂ x≈y) = <-irrefl ((IsEquivalence.sym (IsPreorder.isEquivalence ≤-isPreorder)) x≈y) x<y -- x<y
+
+    ≤-isPartialOrder : IsPartialOrder _≈_ _≤_
+    ≤-isPartialOrder = record { isPreorder = ≤-isPreorder ; antisym = ≤-antisym }
+
+  record HasDecPartialOrder : Set (α ⊔ suc ℓ) where
     infix 4 _≤?_
     field
       hasPartialOrder : HasPartialOrder
-      _≤?_  : Decidable (HasPreorder._≤_ (HasPartialOrder.hasPreorder hasPartialOrder))
+    open HasPartialOrder hasPartialOrder
+    open HasPreorder hasPreorder
+    field
+      _≤?_ : Decidable _≤_
+
+-- conversions
```

Comment:
The comments here and below are very haphazard.

Reply:
I'll remove the comments.